core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java

```diff
@@ -253,22 +252,26 @@ protected final void setEndpoints(Iterable<Endpoint> endpoints) {
         if (!allowEmptyEndpoints && Iterables.isEmpty(endpoints)) {
             return;
         }
-        final List<Endpoint> oldEndpoints = this.endpoints;
-        final List<Endpoint> newEndpoints = ImmutableList.sortedCopyOf(endpoints);
-
-        if (!hasChanges(oldEndpoints, newEndpoints)) {
-            return;
-        }
 
-        endpointsLock.lock();
+        lock();
         try {
-            this.endpoints = newEndpoints;
+            final List<Endpoint> oldEndpoints = this.endpoints;
+            final List<Endpoint> newEndpoints = ImmutableList.sortedCopyOf(endpoints);
+
+            if (!hasChanges(oldEndpoints, newEndpoints)) {
+                return;
+            }
+
+            unsafeUpdateEndpoints(newEndpoints);
         } finally {
-            endpointsLock.unlock();
+            unlock();
         }
+    }
 
+    private void unsafeUpdateEndpoints(List<Endpoint> newEndpoints) {
+        endpoints = newEndpoints;
         notifyListeners(newEndpoints);
```

Comment:
Shouldn't we call `listener.accept(latestValue);` out of the lock?
How about adding `getListeners` to `AbstractListenable` and calling
```java
for (Consumer<? super T> listener : listeners) {
    listener.accept(latestValue);
}
```
out of the lock as is?

Reply:
I'd tried to refactor the code to avoid locking for notification but the code becomes too complex.
```java
protected final void removeEndpoint(Endpoint e) {
    lock();
    boolean unlocked = false;
    try {
        final List<Endpoint> newEndpoints = endpoints.stream()
                                                     .filter(endpoint -> !endpoint.equals(e))
                                                     .collect(toImmutableList());
        updateEndpoints(newEndpoints, true);
        unlocked = true;
    } finally {
        if (!unlocked) {
            unlock();
        }
    }
}

private void updateEndpoints(Iterable<Endpoint> endpoints, boolean locked) {
    if (!locked) {
        lock();
    }

    List<Consumer<List<Endpoint>>> listeners = null;
    List<Endpoint> newEndpoints = null;
    try {
        if (!allowEmptyEndpoints && Iterables.isEmpty(endpoints)) {
            return;
        }

        final List<Endpoint> oldEndpoints = this.endpoints;
        final List<Endpoint> sorted = ImmutableList.sortedCopyOf(endpoints);
        if (!hasChanges(oldEndpoints, sorted)) {
            return;
        }

        this.endpoints = newEndpoints = sorted;
        listeners = listeners();
    } finally {
        unlock();

        if (newEndpoints != null) {
            if (listeners != null) {
                for (Consumer<List<Endpoint>> listener : listeners) {
                    listener.accept(newEndpoints);
                }
            }
            maybeCompleteInitialEndpointsFuture(newEndpoints);
        }
    }
}
```

Are you worried about if a listener acquires a lock for a long time?

Reply:
> Are you worried about if a listener acquires a lock for a long time?

Yes, and I think the current design is based on the same notion. We don't usually call listeners while holding a lock.
But if it's unavoidable, I think we can make an exception for the case.

> t the code becomes too complex.

I was rather thinking something like this:
```java
protected final void setEndpoints(Iterable<Endpoint> endpoints) {
    if (!allowEmptyEndpoints && Iterables.isEmpty(endpoints)) {
        return;
    }

    // Might need to change the type if it's ugly.
    final Consumer<? super List<Endpoint>>[] listeners;
    final List<Endpoint> newEndpoints;
    lock();
    try {
        final List<Endpoint> oldEndpoints = this.endpoints;
        newEndpoints = ImmutableList.sortedCopyOf(endpoints);

        if (!hasChanges(oldEndpoints, newEndpoints)) {
            return;
        }

        this.endpoints = newEndpoints;
        listeners = copyListeners();
    } finally {
        unlock();
    }

    for (Consumer<? super List<Endpoint>> listener : listeners) {
        listener.accept(newEndpoints);
    }
    maybeCompleteInitialEndpointsFuture(newEndpoints);
}
```

Reply:
> We don't usually call listeners while holding a lock.

Q: Should we fix the following code block as well?
https://github.com/line/armeria/blob/496a3f63621b68b73fbd626878c811c2acf3d007/core/src/main/java/com/linecorp/armeria/common/util/AbstractListenable.java#L90-L95